Add support for the LongEmbed benchmark

dwzhu-pku opened the pull request:
<!-- If you are not submitting for a dataset, feel free to remove the content below  -->


<!-- add additional description, question etc. related to the new dataset -->

## Checklist for adding MMTEB dataset

<!-- 
Before you commit here is a checklist you should complete before submitting
if you are not 
 -->
Reason for dataset addition:
<!-- Add reason for adding dataset here. E.g. it covers task/language/domain previously not covered -->


- [x] I have tested that the dataset runs with the `mteb` package.
- [ ] I have run the following models on the task (adding the results to the pr). These can be run using the `mteb run -m {model_name} -t {task_name}` command.
  - [ ] `sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2`
  - [ ] `intfloat/multilingual-e5-small`
- [x] I have checked that the performance is neither trivial (both models gain close to perfect scores) nor random (both models gain close to random scores).
- [x] I have considered the size of the dataset and reduced it if it is too big (2048 examples is typically large enough for most tasks)
- [x] Run tests locally to make sure nothing is broken using `make test`. 
- [x] Run the formatter to format the code using `make lint`. 
- [ ] I have added points for my submission to the [POINTS.md](https://github.com/embeddings-benchmark/mteb/blob/main/docs/mmteb/points.md) file. (i will add these information when the arxiv version of our pape is out)

isaac-chung reviewed (commented):
Could you also add results from running `mteb run -m {model_name} -t {task_name}` please?

KennethEnevoldsen commented:
@isaac-chung, if you want to take charge of some of these reviews that would be great. If so I can add you to the list of reviewers for MMTEB (see [overview](https://github.com/embeddings-benchmark/mteb/issues/360) issue)

KennethEnevoldsen commented on mteb/tasks/Retrieval/eng/LEMBNarrativeQARetrieval.py:
```diff
@@ -0,0 +1,67 @@
+        eval_splits=[_EVAL_SPLIT],
+        eval_langs=["eng-Latn"],
+        main_score="ndcg_at_10",
+        date=None,
+        form=None,
+        domains=None,
+        task_subtypes=None,
+        license=None,
+        socioeconomic_status=None,
+        annotations_creators=None,
+        dialect=None,
+        text_creation=None,
+        bibtex_citation=None,
+        n_samples=None,
+        avg_character_length=None,
```
These need to be filled out. The docstring of the taskmetadata gives you an introduction to each field.

dwzhu-pku replied:
ok, I have filled some key parameters. Personally, I'm confused by `socioeconomic_status`, what do you mean by saying the `socioeconomic_status` of a dataset?

KennethEnevoldsen commented on mteb/tasks/Retrieval/eng/LEMBSummScreenFDRetrieval.py:
```diff
@@ -0,0 +1,84 @@
+        avg_character_length=None,
+    )
+
+    # @property
+    # def description(self):
+    #     return {
+    #         'name': 'LEMBSummScreenFDRetrieval',
+    #         'dataset_name': 'dwzhu/LongEmbed',
+    #         'subset': 'summ_screen_fd',
+    #         'reference': 'https://huggingface.co/datasets/dwzhu/LongEmbed',
+    #         "description": (
+    #             "summ_screen_fd subset of dwzhu/LongEmbed dataset. "
+    #         ),
+    #         "type": "Retrieval",
+    #         "category": "s2p",
+    #         "eval_splits": ["validation"],
+    #         "eval_langs": ["en"],
+    #         "main_score": "ndcg_at_10",
+    #     }
```
```suggestion
```

dwzhu-pku replied:
thanks for this, I have removed these dead codes

KennethEnevoldsen reviewed (changes requested):
Wonderful addition! Seems like we a still missing a bit of metadata as well as model scores.

dwzhu-pku commented:
Thanks to @KennethEnevoldsen @isaac-chung for your timely review! I have updated the PR, where I added some model scores and meta data. Specifically, I have included results on all six datasets except for Passkey and Needle, which requires an additional parameter of `context_length`, and results of different `context_length` just overwrites each other. For the meta data, I have added some key information such as domains and forms. However, I have also find some fields very confusing. For example, what do you mean by saying the `socioeconomic_status` of a dataset?

KennethEnevoldsen commented on mteb/tasks/Retrieval/eng/LEMBNarrativeQARetrieval.py:
```diff
@@ -0,0 +1,67 @@
+        eval_splits=[_EVAL_SPLIT],
+        eval_langs=["eng-Latn"],
+        main_score="ndcg_at_10",
+        date=None,
```
when the text was created (from, to) or best guess.

dwzhu-pku replied:
take Narrative for example, do you mean when the original dataset is created (which is back in 2020), or when it is curated for retrieval (which could be today)

KennethEnevoldsen replied:
Text creation time, so if it e.g. includes books written from 1800-2000. These can be approximate guesses. Given that it was created in 2020, that at least set the upper bound.

dwzhu-pku replied:
I see, thank you!

KennethEnevoldsen commented on mteb/tasks/Retrieval/eng/LEMBNarrativeQARetrieval.py:
```diff
@@ -0,0 +1,67 @@
+        form=["written"],
+        domains=["Fiction", "Non-fiction"],
+        task_subtypes=["Article retrieval"],
+        license=None,
```
required, if not specified then use "Not specified"

dwzhu-pku replied:
ok

KennethEnevoldsen commented on mteb/tasks/Retrieval/eng/LEMBNarrativeQARetrieval.py:
```diff
@@ -0,0 +1,67 @@
+        domains=["Fiction", "Non-fiction"],
+        task_subtypes=["Article retrieval"],
+        license=None,
+        socioeconomic_status=None,
```
The socioeconomic status of the text creators (lawyers is high, social media is mixed, news articles would be medium, but depends on the outlet)

It is just a rough estimate

dwzhu-pku replied:
ok, I see now. So the document here is referring to the text creators, rather than the data itself, right? https://github.com/embeddings-benchmark/mteb/blob/main/mteb/abstasks/TaskMetadata.py#L139

KennethEnevoldsen replied:
Ahh right, yea that is not very clear. Feel free to suggest a reformulation and add a point to (bug)fixes for yourself.

KennethEnevoldsen commented on mteb/tasks/Retrieval/eng/LEMBNarrativeQARetrieval.py:
```diff
@@ -0,0 +1,67 @@
+        text_creation=None,
+        bibtex_citation=None,
+        n_samples={_EVAL_SPLIT: 10449},
+        avg_character_length=None,
```
you can use the `calculate_metadata_metrics` from abstask retrieval for this.

dwzhu-pku replied:
I see. Actually, what confuses me is the document saying that for retrieval task, 'this should be the average character length of the query-document pairs'. What's the average character length of a *pair* ? should I report a pair of average length, or the average length of the concatenation of query and document?

KennethEnevoldsen replied:
I agree it isn't too clear, which is why we specified the function (to at least make it consistent).

KennethEnevoldsen commented on mteb/tasks/Retrieval/eng/LEMBNarrativeQARetrieval.py:
```diff
@@ -0,0 +1,67 @@
+        annotations_creators="derived",
+        dialect=None,
+        text_creation=None,
+        bibtex_citation=None,
```
It seems like it is from a paper so should be specified

dwzhu-pku replied:
ok

KennethEnevoldsen commented on mteb/tasks/Retrieval/eng/LEMBNarrativeQARetrieval.py:
```diff
@@ -0,0 +1,67 @@
+        socioeconomic_status=None,
+        annotations_creators="derived",
+        dialect=None,
+        text_creation=None,
```
How was the text created? E.g. "Found" online

dwzhu-pku replied:
I think it partly overlaps with `annotations_creators`. Isn't it enough to set `annotations_creators`  to `derived`, if I have just transformed existing dataset for retrieval?

KennethEnevoldsen commented on mteb/tasks/Retrieval/eng/LEMBNarrativeQARetrieval.py:
```diff
@@ -0,0 +1,67 @@
+        license=None,
+        socioeconomic_status=None,
+        annotations_creators="derived",
+        dialect=None,
```
assuming there are no dialects:
```suggestion
        dialect=[],
```

dwzhu-pku replied:
great

dwzhu-pku replied:
Thanks a lot for you elaboration! @KennethEnevoldsen

KennethEnevoldsen reviewed (commented):
Tried to give reasonable directions on the metadata. Let us know if the documentation is lacking.

dwzhu-pku commented:
many thanks to @KennethEnevoldsen  for the detailed elaboration on each parameter for meta data. I have filled them all, and added results on Passkey and Needle using 512  context length.

KennethEnevoldsen commented on results/intfloat__multilingual-e5-small/LEMBNeedleRetrieval.json:
```diff
@@ -0,0 +1,43 @@
+    "mrr_at_3": 0.86667,
+    "mrr_at_5": 0.87067,
+    "ndcg_at_1": 0.8,
+    "ndcg_at_10": 0.90614,
```
Seems like performance here is quite high. Does that correspond with your expectations? Not too familiar with the dataset, but it might be worth either using ndcg_at_1 or whatever they use in the paper.

dwzhu-pku replied:
see below.

KennethEnevoldsen commented on results/intfloat__multilingual-e5-small/LEMBPasskeyRetrieval.json:
```diff
@@ -0,0 +1,43 @@
+    "mrr_at_20": 1.0,
+    "mrr_at_3": 1.0,
+    "mrr_at_5": 1.0,
+    "ndcg_at_1": 1.0,
+    "ndcg_at_10": 1.0,
+    "ndcg_at_100": 1.0,
+    "ndcg_at_1000": 1.0,
+    "ndcg_at_20": 1.0,
+    "ndcg_at_3": 1.0,
+    "ndcg_at_5": 1.0,
```
Perfect scores?

dwzhu-pku replied:
yes, it is expected. The passkey and needle test have various evaluation lengths: {256,512,1024,2048,4096,8192,16384,32768}. (That's why we need an additional `context length` parameter) This is designed to measure the effective context window length of the embedding models. The reported scores are from 512 context length. Since embedding models typically have a context length of 512 and above, it's not strange for the perfect score here. After all, passkey retrieval itself is very easy. Needle retrieval is a little bit harder, so we did not observe 100% accuracy here

KennethEnevoldsen replied:
Ahh thanks for clarifying. This seems like they should have their own task subtypes.

A potential solution could simply be to include an equal percentage of each size. Then models with higher context length would perform better. An alternative model. Another option would also be to use the splits to differentiate. So e.g. "test_256", "test_512" etc.?

dwzhu-pku replied:
Actually it also bothers me in implementation. The first solution by *include an equal percentage of each size* may not work, as I do not want the candidate documents for each length to be mixed together. As for the second choice, I have used `split` to specify `corpus`, `queries` and `qrels`. As a result, I cannot figure out a better way then passing the `context_length` parameter to do the filtering.

KennethEnevoldsen replied:
Well the "test" split seems to be used in the data structure (not needed on HF), so splitting it up during the  the dataset_transform() seems like a valid approach.

dwzhu-pku replied:
make sense, I'll try it. Really appreciate your help.

KennethEnevoldsen replied:
No worries, sorry for the long review period, but this seems like a good thing to get right early on in the process.

dwzhu-pku replied:
Hi, I have changed the code to use `test_256`, `test_512`, ... instead of the context length. Looks much better now, really appreciate your reminder! Results are also updated. Would you like to take a look? By the way, we have made our repo publicly available here: https://github.com/dwzhu-pku/LongEmbed. Welcome to take a look!

KennethEnevoldsen reviewed (commented):
Thanks! Seems like there are a few odd scores, we should probably look into, but other than that the PR looks good.

KennethEnevoldsen commented on mteb/tasks/Retrieval/eng/LEMBWikimQARetrieval.py:
```diff
@@ -0,0 +1,78 @@
+        ]  # dict_keys(['qid', 'doc_id'])
+        qrels = {row["qid"]: {row["doc_id"]: 1} for row in qrels_list}
+
+        print("Example Query")
+        print(list(queries.values())[5])
+        print("Example Passage (truncate at 200 characters)")
+        print(list(corpus.values())[5]["text"][:200])
```
```suggestion
```

KennethEnevoldsen commented on mteb/tasks/Retrieval/eng/LEMBSummScreenFDRetrieval.py:
```diff
@@ -0,0 +1,87 @@
+        ]  # dict_keys(['qid', 'doc_id'])
+        qrels = {row["qid"]: {row["doc_id"]: 1} for row in qrels_list}
+
+        print("Example Query")
+        print(list(queries.values())[5])
+        print("Example Passage (truncate at 200 characters)")
+        print(list(corpus.values())[5]["text"][:200])
```
delete

KennethEnevoldsen commented on mteb/tasks/Retrieval/eng/LEMBQMSumRetrieval.py:
```diff
@@ -0,0 +1,100 @@
+        ]  # dict_keys(['qid', 'doc_id'])
+        qrels = {row["qid"]: {row["doc_id"]: 1} for row in qrels_list}
+
+        print("Example Query")
+        print(list(queries.values())[5])
+        print("Example Passage (truncate at 200 characters)")
+        print(list(corpus.values())[5]["text"][:200])
```
delete

KennethEnevoldsen commented on mteb/tasks/Retrieval/eng/LEMBPasskeyRetrieval.py:
```diff
@@ -0,0 +1,76 @@
+        qrels_list = qrels_list.filter(lambda x: x["context_length"] == context_length)
+        qrels = {row["qid"]: {row["doc_id"]: 1} for row in qrels_list}
+
+        print("Example Query")
+        print(list(queries.values())[5])
+        print("Example Passage (truncate at 200 characters)")
+        print(list(corpus.values())[5]["text"][:200])
```
delete

KennethEnevoldsen commented on mteb/tasks/Retrieval/eng/LEMBNeedleRetrieval.py:
```diff
@@ -0,0 +1,76 @@
+        qrels_list = qrels_list.filter(lambda x: x["context_length"] == context_length)
+        qrels = {row["qid"]: {row["doc_id"]: 1} for row in qrels_list}
+
+        print("Example Query")
+        print(list(queries.values())[5])
+        print("Example Passage (truncate at 200 characters)")
+        print(list(corpus.values())[5]["text"][:200])
```
delete

KennethEnevoldsen commented on mteb/tasks/Retrieval/eng/LEMBNarrativeQARetrieval.py:
```diff
@@ -0,0 +1,91 @@
+        ]  # dict_keys(['qid', 'doc_id'])
+        qrels = {row["qid"]: {row["doc_id"]: 1} for row in qrels_list}
+
+        print("Example Query")
+        print(list(queries.values())[5])
+        print("Example Passage (truncate at 200 characters)")
+        print(list(corpus.values())[5]["text"][:200])
```
delete

dwzhu-pku replied:
ok, let me do the removal

KennethEnevoldsen reviewed (commented):
discovered a few prints which I would remove.

Muennighoff commented:
This looks amazing! If you also want a leaderboard tab for your benchmark feel free to open a PR here: https://huggingface.co/datasets/mteb/results with the result files. We could add it under Retrieval similar to https://huggingface.co/spaces/mteb/leaderboard?task=retrieval&language=law

dwzhu-pku commented:
> This looks amazing! If you also want a leaderboard tab for your benchmark feel free to open a PR here: https://huggingface.co/datasets/mteb/results with the result files. We could add it under Retrieval similar to https://huggingface.co/spaces/mteb/leaderboard?task=retrieval&language=law

Thanks!

KennethEnevoldsen commented on mteb/tasks/Retrieval/eng/LEMBNeedleRetrieval.py:
```diff
@@ -0,0 +1,104 @@
+        annotations_creators="derived",
+        dialect=[],
+        text_creation="found",
+        bibtex_citation=None,
```
Missing?

dwzhu-pku replied:
these are two synthetic tasks, I guess it's not necessary to provide citations, as different people can synthetic different data?

dwzhu-pku replied:
if must cite, I think I should cite my paper, which is on hold on arxiv not ...

KennethEnevoldsen replied:
I would cite the paper that introduces the synthetic task - sounds like that it your paper (in which case I would add that).

dwzhu-pku replied:
thanks for this! Since its publication has been delayed (its status is `on hold` in arxiv, which I have no idea will take how long), can I add these fields later?

KennethEnevoldsen replied:
Sure let us just get this merged in

dwzhu-pku replied:
Thanks so much to you guys!

KennethEnevoldsen commented on mteb/tasks/Retrieval/eng/LEMBPasskeyRetrieval.py:
```diff
@@ -0,0 +1,104 @@
+        annotations_creators="derived",
+        dialect=[],
+        text_creation="found",
+        bibtex_citation=None,
```
missing

KennethEnevoldsen reviewed (approved):
This looks really good, very happy with how it has ended up. @Muennighoff as you suggest let us add a page on the

- I found two missing citations. 
- Will you also add the points as well. 
- We can create the benchmark in a separate PR (this one has already gone on for a while now)

KennethEnevoldsen commented:
merging into branch to add points and then merge

tomaarsen commented:
Excellent work!

- Tom Aarsen

dwzhu-pku commented:
hi, I'm ready to add points for this submission, may I ask what kind of information should I provide? I think I'm still a little bit confused about what to fill in these fields: @KennethEnevoldsen 
```
{
    "GitHub": "GitHubUser1",
    "New dataset": 2-6,  # 2 points for the dataset and 4 points for the task
    "New task": 2, # e.g. a new style of task (e.g. classification, or retrieval)
    "Dataset annotations": 1, # 1 point for each full dataset annotation
    "Bug fixes": 2-10, # depends on the complexity of the fix
    "Running Models": 1, # pr model run
    "Review PR": 2, # two points pr. reviewer, can be given to multiple reviewers
    "Paper Writing": NA, 
    "Ideation": NA,
    "Coordination": NA
}
```
